File: js-port/funcscript-js/src/funcs/misc/error-function.js
const { BaseFunction, CallType } = require('../../core/function-base');
const {
  ensureTyped,
  typeOf,
  valueOf
} = require('../../core/value');
const { FSDataType } = require('../../core/fstypes');
const helpers = require('../helpers');

class ErrorFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'error';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    if (parameters.count < 1 || parameters.count > 2) {
      return helpers.makeError(
        helpers.FsError.ERROR_PARAMETER_COUNT_MISMATCH,
        `${this.symbol}: message and optional type expected`
      );
    }

    const messageTyped = ensureTyped(parameters.getParameter(provider, 0));
    if (typeOf(messageTyped) !== FSDataType.String) {
      return helpers.makeError(
        helpers.FsError.ERROR_TYPE_MISMATCH,
        `${this.symbol}: message must be a string`
      );
    }

    let typeValue = helpers.FsError.ERROR_DEFAULT;
    if (parameters.count > 1) {
      const typeTyped = ensureTyped(parameters.getParameter(provider, 1));
      const typeKind = typeOf(typeTyped);
      if (typeKind === FSDataType.Null) {
        typeValue = helpers.FsError.ERROR_DEFAULT;
      } else if (typeKind === FSDataType.String) {
        const raw = valueOf(typeTyped);
        typeValue = raw && raw.length > 0 ? raw : helpers.FsError.ERROR_DEFAULT;
      } else {
        return helpers.makeError(
          helpers.FsError.ERROR_TYPE_MISMATCH,
          `${this.symbol}: optional type must be a string`
        );
      }
    }

    const message = valueOf(messageTyped);
    return helpers.makeError(typeValue, message);
  }
}

module.exports = {
  ErrorFunction
};

File: FuncScript/Functions/Misc/ErrorFunction.cs
using FuncScript.Core;
using FuncScript.Model;

namespace FuncScript.Functions.Misc
{
    public class ErrorFunction : IFsFunction
    {
        public const string SYMBOL = "error";

        public int MaxParsCount => 2;

        public CallType CallType => CallType.Prefix;

        public string Symbol => SYMBOL;

        public int Precedence => 0;

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (pars.Count < 1 || pars.Count > MaxParsCount)
                throw new Error.TypeMismatchError($"{this.Symbol}: message and optional type expected");

            var messageValue = pars.GetParameter(parent, 0);
            if (messageValue is not string message)
                throw new Error.TypeMismatchError($"{this.Symbol}: message must be a string");

            string type = null;
            if (pars.Count > 1)
            {
                var typeValue = pars.GetParameter(parent, 1);
                if (typeValue == null)
                    type = null;
                else if (typeValue is string typeString)
                    type = typeString;
                else
                    throw new Error.TypeMismatchError($"{this.Symbol}: optional type must be a string");
            }

            if (string.IsNullOrEmpty(type))
                return new FsError(message);

            return new FsError(type, message);
        }

        public string ParName(int index)
        {
            return index switch
            {
                0 => "Message",
                1 => "ErrorType",
                _ => ""
            };
        }
    }
}
